File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/cs/hello_dynamic/src/HelloPublisher.cs
/* ****************************************************************************
 *         (c) Copyright, Real-Time Innovations, All rights reserved.       
 *                                                                          
 *         Permission to modify and use for internal purposes granted.      
 * This software is provided "as is", without warranty, express or implied. 
 *                                                                          
 * ****************************************************************************
 */

using System;
using System.Collections.Generic;
using System.Text;

namespace Hello {
    /**
     * Manage the data publication for the Hello example 
     */
    class HelloPublisher {
        /* Maximum consecutive write error after stopping the write process */
        private const int MAX_CONSECUTIVE_WRITE_ERROR = 5;

        private DDS.DynamicDataWriter _theDataWriter;
        private DDS.DynamicData _theInstance;
        private int _theVerbose;
        private int _theSampleCount;

        
        // ************************************************************************
        /**
         * Builds a new HelloPublisher object and initialize the inner DDS 
         * entities.
         * This method only creates the entities, but it does not actually write
         * the data.
         * To write the data, use the method writeSamples().
         *  
         * @param participant the DDS Domain participant
         * @param topic       the DDS topic 
         * @param verbose     an integer to control the output verbosity
         * @param dataSize    the size of the payload data
         */

        public HelloPublisher(DDS.DomainParticipant participant,
                              DDS.Topic topic,
                              DDS.TypeCode typeCode,
                              int verbose,
                              int dataSize,
                              int sampleCount) {
            _theVerbose = verbose;
            _theSampleCount = sampleCount;

            /* Creates the DDS Data writer. 
             * Just like before, if you want to customize the writer QoS,
             * use DDS.Publisher.get_default_datawriter_qos() to 
             * initialize a local copy of the default QoS, modify them, then
             * use them in the creation call below instead of 
             * DDS_DATAWRITER_QOS_DEFAULT.
             * For more data writer API info, see:
             *     $NDDSHOME/doc/html/api_dotnet/group__DDSWriterModule.html
             */
            if (verbose > 0) {
                Console.WriteLine("Creating the data writer...");
            }
            _theDataWriter = (DDS.DynamicDataWriter)
                participant.create_datawriter(topic,
                            DDS.Publisher.DATAWRITER_QOS_DEFAULT,
                            null, /* Listener */
                            DDS.StatusMask.STATUS_MASK_NONE);
            if (_theDataWriter == null) {
                Console.Error.WriteLine("! Unable to create DDS data writer");
                return;
            }

            /* Creates an instance of the sparse data we are about to send
             */
            _theInstance = new DDS.DynamicData(typeCode, DDS.DynamicData.DYNAMIC_DATA_PROPERTY_DEFAULT);

            /* The DynamicDataType has been defined (in HelloWorldType.cxx) to hold
             * a structure like the following one:
             * struct HelloWorld {
             *     string<64>            prefix;
             *     long                  sampleId;
             *     sequence<octet, 8192> payload;
             * };
             */
            _theInstance.set_string("prefix",
                                DDS.DynamicData.MEMBER_ID_UNSPECIFIED,
                                "HelloWorld!");
            _theInstance.set_int("sampleId", DDS.DynamicData.MEMBER_ID_UNSPECIFIED,
                                0);
            
            DDS.ByteSeq byteSeq = new DDS.ByteSeq(dataSize);
            //byte[] payload = new byte[dataSize];
            for (int i = 0; i < dataSize; ++i) {
                byteSeq.buffer[i] = (byte)(i % 0xff);
            }
            
            _theInstance.set_byte_seq("payload",
                                DDS.DynamicData.MEMBER_ID_UNSPECIFIED,
                                byteSeq);
        }
        /**
         * Writes samples to DDS. This method never returns.
         */
        public void writeSamples() {
            DDS.InstanceHandle_t instance_handle = DDS.InstanceHandle_t.HANDLE_NIL;
            /* For a data type that has a key, if the same instance is going to be
             * written multiple times, initialize the key here
             * and register the keyed instance prior to writing
             */
            /* instance_handle = writer.register_instance(instance);
             */
            Console.WriteLine("Sending data...");
            int consecutiveErrors = 0;
            for (int count = 0; ; ++count) {
                _theInstance.set_int("sampleId",
                        DDS.DynamicData.MEMBER_ID_UNSPECIFIED,
                        count);
                try {
                    _theDataWriter.write(_theInstance, ref instance_handle);
                    /* Always clear the error count in case of successful write */
                    consecutiveErrors = 0;
                }
                catch (DDS.Retcode_Error e)
                {
                    Console.Error.WriteLine("! Write error " +
                                e.GetType().ToString() + ": " + e.Message);
                    if (++consecutiveErrors > MAX_CONSECUTIVE_WRITE_ERROR) {
                        Console.WriteLine("! Reached maximum number of failure, stopping writer...");
                        return;
                    }
                }
                if (_theVerbose > 0 && (count % 10000) == 0) {
                    Console.WriteLine("Sent " + count + " samples...");
                }
                if (_theSampleCount != 0 && (count >= _theSampleCount)) {
                    Console.WriteLine("Sent " + count + " samples.");
                    break;
                }
            }
        }
    }
}

File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/c++/hello_dynamic/src/HelloPublisher.cxx
/*****************************************************************************/
/*         (c) Copyright, Real-Time Innovations, All rights reserved.        */
/*                                                                           */
/*         Permission to modify and use for internal purposes granted.       */
/* This software is provided "as is", without warranty, express or implied.  */
/*                                                                           */
/*****************************************************************************/

#include <iostream>
#include <string>
#include <cstdlib>
#include "Hello.h"
#include "HelloPublisher.h"

/* Defines the maximum number of consecutive write errors. */
#define MAX_CONSECUTIVE_WRITE_ERROR         10



/*****************************************************************************/
/* start_publisher                                                           */
/*                                                                           */
/* Creates the DDS publisher and data writers, then start sending the data.  */
/*                                                                           */
/* Input:                                                                    */
/*   participant: the DDS domain participant to use                          */
/*   topic: the DDS topic to use for publishing                              */
/*   dataSise: size of the payload buffer to send                            */
/*                                                                           */
/* Returns:                                                                  */
/*   The boolean value true if an error occurred.                            */
/*                                                                           */
/*****************************************************************************/
bool startPublisher(
                        DDSDomainParticipant *participant,
                        DDSTopic *topic,
                        DDSDynamicDataTypeSupport *typeSupport,
                        DDS_Long verbose,
                        DDS_Long dataSize,
                        DDS_Long sampleCount) {
    
    DDSDataWriter *dataWriter            = NULL;
    DDSDynamicDataWriter *helloWriter    = NULL;
    DDS_DynamicData *instance            = NULL;
    bool returnValue                     = false;
    DDS_Long i, count;
    DDS_ReturnCode_t rc;
    DDS_OctetSeq payloadSequence;
    DDS_Duration_t disc_period = {1, 0};    

    /* Creates the DDS Data writer. 
     * Just like before, if you want to customize the writer QoS,
     * use DDSPublisher::get_default_datawriter_qos() to 
     * initialize a local copy of the default QoS, modify them, then
     * use them in the creation call below instead of 
     * DDS_DATAWRITER_QOS_DEFAULT.
     * For more data writer API info, see:
     *     $NDDSHOME/doc/html/api_cpp/group__DDSWriterModule.html
     */
    if (verbose) {
        std::cout << "Creating the data writer..." << std::endl;
    }
    dataWriter = participant->create_datawriter(
                        topic,
                        DDS_DATAWRITER_QOS_DEFAULT,
                        NULL,           /* listener */
                        DDS_STATUS_MASK_NONE);
    if (dataWriter == NULL) {
        std::cerr << "! Unable to create DDS data writer" << std::endl;
        goto exitFn;
    }

    // wait for discovery
    NDDS_Utility_sleep(&disc_period);

    /* The following narrow function should never fail, as it performs 
     * only a safe cast of the generic data writer into a specific
     * DDSStringDataWriter.
     * In our case the following cast would do the same job:
     *
          helloWriter = static_cast<DDSStringDataWriter *>(dataWriter);
     */
    helloWriter = DDSDynamicDataWriter::narrow(dataWriter);
    if (helloWriter == NULL) {
        std::cerr << "! Unable to narrow data writer into DDSDynamicDataWriter"
                  << std::endl;
        goto exitFn;
    }

    /* Creates an instance of the sparse data we are about to send
     */
    instance = typeSupport->create_data();
    if (instance == NULL) {
        std::cerr << "! Unable to create an instance of the data" << std::endl;
        std::cerr << "! This problem most likely is caused by out of memory"
                  << std::endl;
        goto exitFn;
    }

    /* The DynamicDataType has been defined (in HelloWorldType.cxx) to hold
     * a structure like the following one:
     * struct HelloWorld {
     *     string<64>            prefix;
     *     long                  sampleId;
     *     sequence<octet, 8192> payload;
     * };
     */
    instance->set_string("prefix", 
                        DDS_DYNAMIC_DATA_MEMBER_ID_UNSPECIFIED, 
                        "Hello World!");
    instance->set_long("sampleId", 
                        DDS_DYNAMIC_DATA_MEMBER_ID_UNSPECIFIED, 
                        0);
    payloadSequence.ensure_length(dataSize, dataSize);
    for (i = 0; i < dataSize; ++i) {
        payloadSequence[i] = (i % 0xff);
    }
    instance->set_octet_seq("payload",
                        DDS_DYNAMIC_DATA_MEMBER_ID_UNSPECIFIED, 
                        payloadSequence);
    
    // Send the data!
    std::cout << "Sending data..." << std::endl;
    i = 0; /* Use 'i' to keep track of consecutive write errors */
    for (count = 0;; ++count) {
        rc = instance->set_long("sampleId", 
                        DDS_DYNAMIC_DATA_MEMBER_ID_UNSPECIFIED, 
                        count);
        if (rc != DDS_RETCODE_OK) {
            std::cerr << "! Unable to set value (" << count
                      << ") for sampleId. Error=" << rc << std::endl;
            continue;
        }
        rc = helloWriter->write(
                        *instance, 
                        DDS_HANDLE_NIL);
        if (rc != DDS_RETCODE_OK) {
            std::cerr << "! Write error " <<  rc << std::endl;
            if (++i > MAX_CONSECUTIVE_WRITE_ERROR) {
                std::cerr << "! Reached maximum number of failure, "
                          << "stopping writer..." << std::endl;
                goto exitFn;
            }
        } else {
            i = 0;  /* Always clear the error count in case of successful write */
        }
        if (verbose && (count % 10000) == 0) {
            std::cout << "Sent " << count
                      << " samples..." << std::endl;
        }
        if (sampleCount != 0 && (count >= sampleCount)) {
            std::cout << "\nSent " << count
                      << " samples." << std::endl;
            break;
        }
    }
    
    /* This code is reached if sampleCount is set */
    returnValue = true;

exitFn:
    if (instance != NULL) {
        rc = typeSupport->delete_data(instance);
        if (rc != DDS_RETCODE_OK) {
            std::cerr << "! Unable to delete instance data: " 
                      << rc << std::endl;
        }
        instance = NULL;
    }
    
    return returnValue;
}

